update tidb_opt_fix_control for v6.5 (#15997)
pingcap/docs: optimizer-fix-controls.md

---
title: Optimizer Fix Controls
summary: Learn about the Optimizer Fix Controls feature and how to use `tidb_opt_fix_control` to control the TiDB optimizer in a more fine-grained way.
---

# Optimizer Fix Controls

As the product evolves iteratively, the behavior of the TiDB optimizer changes, which in turn generates more reasonable execution plans. However, in some particular scenarios, the new behavior might lead to unexpected results. For example:

- The effect of some behaviors relies on a specific scenario. Changes that bring improvements to most scenarios might cause regressions to others.
- Sometimes, the relationship between changes in the behavior details and their consequences is very complicated. An improvement in a certain behavior might cause overall regression.

Therefore, TiDB provides the Optimizer Fix Controls feature that allows you to make fine-grained control of TiDB optimizer behaviors by setting values for a group of fixes. This document describes the Optimizer Fix Controls feature and how to use them, and lists all the fixes that TiDB currently supports for Optimizer Fix Controls.

## Introduction to `tidb_opt_fix_control`

> Starting from v6.5.7, TiDB provides the [`tidb_opt_fix_control`](/system-variables.md#tidb_opt_fix_control-new-in-v657) system variable to control the behavior of the optimizer in a more fine-grained way.

Reviewer: ```suggestion
Starting from v6.5.3, TiDB provides the [`tidb_opt_fix_control`](/system-variables.md#tidb_opt_fix_control-new-in-v653) system variable to control the behavior of the optimizer in a more fine-grained way.
```

> Starting from v7.1.0, TiDB provides the [`tidb_opt_fix_control`](/system-variables.md#tidb_opt_fix_control-new-in-v657-and-v710) system variable to control the behavior of the optimizer in a more fine-grained way.

Reviewer: ```suggestion
Starting from v6.5.7, TiDB provides the [`tidb_opt_fix_control`](/system-variables.md#tidb_opt_fix_control-new-in-v657) system variable to control the behavior of the optimizer in a more fine-grained way.
```

Each fix is a control item used to adjust the behavior in the TiDB optimizer for one particular purpose. It is denoted by a number that corresponds to a GitHub Issue that contains the technical details of the behavior change. For example, for fix `44262`, you can review what it controls in [Issue 44262](https://github.com/pingcap/tidb/issues/44262).

> The [`tidb_opt_fix_control`](/system-variables.md#tidb_opt_fix_control-new-in-v657-and-v710) system variable accepts multiple fixes as one value, separated by commas (`,`). The format is `"<#issue1>:<value1>,<#issue2>:<value2>,...,<#issueN>:<valueN>"`, where `<#issueN>` is the fix number. For example:

Reviewer: ```suggestion
The [`tidb_opt_fix_control`](/system-variables.md#tidb_opt_fix_control-new-in-v653) system variable accepts multiple fixes as one value, separated by commas (`,`). The format is `"<#issue1>:<value1>,<#issue2>:<value2>,...,<#issueN>:<valueN>"`, where `<#issueN>` is the fix number. For example:
```

```sql
SET SESSION tidb_opt_fix_control = '44262:ON,44389:ON';
```

## Optimizer Fix Controls reference

> ### [`44262`](https://github.com/pingcap/tidb/issues/44262) <span class="version-mark">New in v6.5.7</span>

Reviewer: ```suggestion
### [`44262`](https://github.com/pingcap/tidb/issues/44262) <span class="version-mark">New in v6.5.3</span>
```

> ### [`44262`](https://github.com/pingcap/tidb/issues/44262) <span class="version-mark">New in v7.2.0</span>

Reviewer: ```suggestion
### [`44262`](https://github.com/pingcap/tidb/issues/44262) <span class="version-mark">New in v6.5.7</span>
```

- Default value: `OFF`
- Possible values: `ON`, `OFF`
- This variable controls whether to allow the use of [Dynamic pruning mode](/partitioned-table.md#dynamic-pruning-mode) to access the partitioned table when the [GlobalStats](/statistics.md#collect-statistics-of-partitioned-tables-in-dynamic-pruning-mode) are missing.

> ### [`44389`](https://github.com/pingcap/tidb/issues/44389) <span class="version-mark">New in v6.5.7</span>

Reviewer: ```suggestion
### [`44389`](https://github.com/pingcap/tidb/issues/44389) <span class="version-mark">New in v6.5.3</span>
```

> ### [`44389`](https://github.com/pingcap/tidb/issues/44389) <span class="version-mark">New in v7.2.0</span>

Reviewer: ```suggestion
### [`44389`](https://github.com/pingcap/tidb/issues/44389) <span class="version-mark">New in v6.5.7</span>
```

- Default value: `OFF`
- Possible values: `ON`, `OFF`
- For filters such as `c = 10 and (a = 'xx' or (a = 'kk' and b = 1))`, this variable controls whether to try to build more comprehensive scan ranges for `IndexRangeScan`.

> ### [`44830`](https://github.com/pingcap/tidb/issues/44830) <span class="version-mark">New in v7.3.0</span>

Reviewer: ```suggestion
### [`44830`](https://github.com/pingcap/tidb/issues/44830) <span class="version-mark">New in v6.5.7</span>
```

- Default value: `OFF`
- Possible values: `ON`, `OFF`
- This variable controls whether Plan Cache is allowed to cache execution plans with the `PointGet` operator generated during physical optimization.

> ### [`44855`](https://github.com/pingcap/tidb/issues/44855) <span class="version-mark">New in v6.5.7</span>

Reviewer: ```suggestion
### [`44855`](https://github.com/pingcap/tidb/issues/44855) <span class="version-mark">New in v6.5.4</span>
```

> ### [`44855`](https://github.com/pingcap/tidb/issues/44855) <span class="version-mark">New in v7.3.0</span>

Reviewer: ```suggestion
### [`44855`](https://github.com/pingcap/tidb/issues/44855) <span class="version-mark">New in v6.5.7</span>
```
